### PR TITLE
feat: support plan generating & execution for new DDL & DML design

### PR DESCRIPTION
**Pull request by xx01cyx**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR applies `SourceExecutorV2`, `DmlExecutor`, `RowIdGenExecutor`, and `DmlManager` to query execution. 
For example, the query plan for `CREATE TABLE t (v int)` will be:

```SQL
 StreamMaterialize { columns: [v, _row_id(hidden)], pk_columns: [_row_id] }
 └─StreamExchange { dist: HashShard(_row_id) }
   └─StreamRowIdGen { row_id_index: 1 }
     └─StreamDml { columns: [v, _row_id] }
       └─StreamSource
```

Some explanations:
- `StreamSource` here contains no actual external streaming source. It is only responsible for receiving barriers.
- `StreamDml` will receive data from `InsertExecutor`, `DeleteExecutor`, and `UpdateExecutor`.
- `StreamRowIdGen` will generate row id for the data. In this case, the primary key is not defined by the user, so we internally add a `_row_id` column as the primary key. If the table has a user-defined primary key, then this executor can be eliminated.

Note that now **"source" stands for streaming source only**. There is **NO table source** now. Though `CREATE TABLE` will create a `StreamSource`, it actually contains nothing related to a source (catalog).

## Detailed Explanations

The following sections describe the modifications on different modules, and might serve as a guide for code review. 

### Batch

In `InsertExecutor`, `DeleteExecutor`, and `UpdateExecutor`, **use `DmlManager`** instead of `TableSourceManager`. The executors get a `sender` (or writer) from `DmlManager`, where the `receiver` is held by `DmlExecutor`. 

#### Binder

- Modify `BoundInsert`, `BoundDelete`, and `BoundUpdate` to hold necessary information for planning.
- Add `resolve_dml_table` method to validate the data manipulation operation.

#### Catalog

- **Source Catalog**
   Remove `SourceKind` because now source catalog is for streaming source only. There is no table source now.

#### Handler

- **`create_source`**
   Not change much. Still retain `CREATE MATERIALIZED SOURCE` until the next release.

- **`create_table`**
   - Move the logic of `gen_materialized_plan` into `gen_create_table_plan`. Will remove `gen_materialized_plan` after unsupporting `CREATE MATERIALIZED SOURCE`.
   - When generating query plan for DDL, use `SourceExecutorV2`, `DmlExecutor`, and`RowIdGenExecutor`.

- **`drop_table`, `drop_source`, `drop_mv`, `drop_index`**
   Since we no longer have table source, the logic for checking existence is modified accordingly.

#### Plan Node

- **Insert, Update, and Delete**
   Add necessary fields the batch executors need in plan nodes.

- **Source**
   Store `column_descs` and `pk_col_ids` in plan node because we need this information for planning even if `source_catalog` is `None` (i.e. no external streaming source).

#### DDL Service

Rename `xxxx_materialized_source` to `xxxx_table`. Make `source` an optional parameter for `create_table`.

#### Stream Graph

Fill in the table id for DML executor when building actor graph.

### RPC Client

Rename `xxxx_materialized_source` to `xxxx_table`.

### Source

Remove `TableSourceManager`, as it is substituted by `DmlManager`.

#### Source Executor

- Remove the old source executor. 
- Add a new field `_store` in `SourceExecutorV2`. This is actually a `FIXME` because it does nothing other than telling rustc the concrete type when constructing `SourceExecutorV2`. (See `SourceExecutorV2::new`)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link

#5949

### CODE REVIEW

**Comment by BugenZhao** [comment]
> * Add a new field `_store` in `SourceExecutorV2`. This is actually a `FIXME` because it does nothing other than telling rustc the concrete type when constructing `SourceExecutorV2`. (See `SourceExecutorV2::new`)

Try try this?

```rust
async fn new_boxed_executor<S: StateStore>(
    params: ExecutorParams,
    node: &Self::Node,
    store: S,
    stream: &mut LocalStreamManagerCore,
) -> StreamResult<BoxedExecutor> {
    ..
    SourceExecutor::<S>::new(.. None, ..)
}
```

**Comment by yezizp2012** [comment]
The mergify will use the content in the "What's changed and what's your intention?" section as the commit message after #6829 . How about simplifying it and put the change details into anther section? So that the commit message would not be too big long. 🥵

**Comment by BugenZhao** [comment]
> The mergify will use the content in the "What's changed and what's your intention?" section as the commit message after #6829. How about simplifying it and put the change details into anther section? So that the commit message would not be too long. 🥵

Should not be a big problem. 🥵 It's only used as the body which is hidden most of the time.

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM. Great job!

File: src/frontend/planner_test/tests/testdata/struct_query.yaml
Comment (xx01cyx) [reply src/frontend/planner_test/tests/testdata/struct_query.yaml:336]: For now, it is invalid to insert into a materialized source. Previously we have no check for this in our code. This PR adds some checks for this (disallow inserting into a materialized source). This test case will thereby fail, so I have to remove it temporarily. I'll add it back after supporting `create table with (connector = '...')`

File: src/meta/src/rpc/service/ddl_service.rs
Comment (xx01cyx) [reply src/meta/src/rpc/service/ddl_service.rs:655]: They are almost the same, except that `create table` doesn't have any internal table. I think it's okay to reuse this method, since `create table` will simply skip iterating over `internal_tables` in this method.

File: src/meta/src/rpc/service/ddl_service.rs
Comment (xx01cyx) [reply src/meta/src/rpc/service/ddl_service.rs:687]: IIUC, if you mean something like: `create table as select * from source`, this should be done by `create materialized view as select * from source`, and this will go to `create_materialized_view` instead of here (`create_table`).

cc @st1page

Comment (BugenZhao) [reply src/meta/src/rpc/service/ddl_service.rs:687]: I mean, is it possible to `CREATE SOURCE source WITH (..)` and then create a DML-able table on top of this `source`, without specifying the properties again? 🤔

File: src/meta/src/stream/stream_graph.rs
Comment (xx01cyx) [reply src/meta/src/stream/stream_graph.rs:571]: Yeah, but in other match arms, for example:

```rust
NodeBody::Arrange(node) => {
    if let Some(table) = &mut node.table {
        update_table(table, "ArrangeNode");
    }
}
```

Just to keep consistent 🤣

File: src/meta/src/rpc/service/ddl_service.rs
Comment (yezizp2012) [reply src/meta/src/rpc/service/ddl_service.rs:650]: It's okay for me to reuse it, would be better to add some comment.

File: src/meta/src/rpc/service/ddl_service.rs
Comment (yezizp2012) [reply src/meta/src/rpc/service/ddl_service.rs:682]: > I mean, is it possible to `CREATE SOURCE source WITH (..)` and then create a DML-able table on top of this `source`, without specifying the properties again? 🤔

I think it's possible and we should support it from the user's perspective. In this case we don't have to fill the source id if its relied source already created.

File: src/source/src/dml_manager.rs
Comment (yezizp2012) [reply src/source/src/dml_manager.rs:66]: > Do we have mechanisms for cleaning unused `TableSourceRef` (may need to be renamed 😄)? For example, the weak references in previous table source manager.

+1, we need to clean the entry when table/source dropped.

**Review by yezizp2012 — COMMENTED** [review]
Great job!!! Generally LGTM.

**Review by st1page — COMMENTED** [review]
great work! LGTM, but the changes of the PR is big and is a part of the feature "create table with connector". Maybe Postponing the PR merging to the next release is a better choice? just one day.

**Comment by xx01cyx** [comment]
> great work! LGTM, but the changes of the PR is big and is a part of the feature "create table with connector". Maybe Postponing the PR merging to the next release is a better choice? just one day.

Sure. Let's wait till next week.

**Review by st1page — APPROVED** [review]

File: src/frontend/planner_test/tests/testdata/struct_query.yaml
Comment (BugenZhao) [reply src/frontend/planner_test/tests/testdata/struct_query.yaml:336]: > insert into a materialized source

Interesting. 😄

File: src/source/src/dml_manager.rs
Comment (BugenZhao) [inline src/source/src/dml_manager.rs:86]: How about making this message more specific? Like "all readers have been dropped".

File: src/source/src/dml_manager.rs
Comment (BugenZhao) [inline src/source/src/dml_manager.rs:47]: According to the method name, how about directly returning the `TableStreamReader`?

File: src/source/src/dml_manager.rs
Comment (BugenZhao) [inline src/source/src/dml_manager.rs:54]: Is there any way to clean up the invalid weak ref entries?

File: src/frontend/src/test_utils.rs
Comment (yezizp2012) [reply src/frontend/src/test_utils.rs:239]: Yes, took a quick look at current implementation, `optional_associated_source_id` is only used by `drop_materialized_source`(renamed to `drop_table_with_source`) to validate the parameter `source_id` passed from FE. I think we can remove it without doing anything else. The `source` and `table` should be create or drop synchronously in this scenario. And for others like creating table from existing source, we can leave it to the "dependent relation".

File: src/source/src/dml_manager.rs
Comment (BugenZhao) [reply src/source/src/dml_manager.rs:54]: https://github.com/risingwavelabs/risingwave/blob/820ff8c95be38469d5757751962f92046d299f98/src/source/src/manager.rs#L138 for example

File: src/source/src/dml_manager.rs
Comment (xx01cyx) [reply src/source/src/dml_manager.rs:47]: If we directly return `TableStreamReader` here, then no one will get hold of the strong ref of `TableSource`. It will thus be dropped immediately because only `DmlManager` has a weak ref of it. We might have to refactor `TableSource` later.

Comment (BugenZhao) [reply src/source/src/dml_manager.rs:47]: Yes. Makes sense.

**Review by BugenZhao — APPROVED** [review]
LGTM! 🚀🚀

**Review by yezizp2012 — APPROVED** [review]
LGTM!! Great job!

**Review by BugenZhao — COMMENTED** [review]
Please remove the mock data in `dashboard` and `package-lock.json` before getting merged.

**Comment by xx01cyx** [comment]
> Please remove the mock data in dashboard and package-lock.json before getting merged.

Seems that I accidentally got this committed 😥

**Pull request merged**
